### PR TITLE
Add a programmatic API (fixes #86)

### LINKED ISSUE #86: Programmatic API
For loading `serve` inside an app

### PR DESCRIPTION
**Pull request by dar5hak**
This is a prototype of `api.js` as discussed in #86. Here's what I think it should do:

- [x] Export a function that takes `options` and `directory` as arguments
- [x] Start a server based on those
- [x] Print the server info nicely, with colours and stuff (tested on Linux)
- [x] Take `stdin`, such as <kbd>^C</kbd> and behave accordingly
- [x] Return something to the caller (what?)
- [x] Write docs

### CODE REVIEW

**Comment by dar5hak** [comment]
Do you think this list is comprehensive? I'm still trying to figure out 4. Any hints on 5?

**Comment by leo** [comment]
@dar5hak Please return an object with a `.stop()` method. This will solve both 4 and 5 as it allows people to stop the server programmatically. 🙏

File: lib/api.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// Native
+const {spawn} = require('child_process')
+const path = require('path')
+
+// { port: 1337 } => ['--port', '1337']
+const optionsToArgs = options => {
```
Comment (tunnckoCore) [inline lib/api.js]: can use `dargs` directly

**Comment by dar5hak** [comment]
I think it covers everything now.

@leo Could you have a final look?

**Comment by dar5hak** [comment]
I'll push the docs tomorrow. Added that to the checklist.

File: lib/api.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+const {spawn} = require('child_process')
+const path = require('path')
+
+// Packages
+const dargs = require('dargs')
+
+module.exports = (options = {}, directory = process.cwd()) => {
+  const scriptPath = path.join(__dirname, '..', 'bin', 'serve.js')
+  const aliases = {cors: 'o'}
+  options._ = [directory]  // Let dargs handle the directory argument
```
Comment (leo) [inline lib/api.js]: Please add an empty line before this line.

File: lib/api.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  const cli = spawn('node',
```
Comment (leo) [inline lib/api.js]: Please split it into multiple lines:

```js
const args = [
  scriptPath,
  ...dargs(options, {aliases})
]

const cli = spawn('node', args, {
  stdio: 'inherit'
})
```

File: package.json [hunk 4]
```diff
@@ -53,6 +54,7 @@
     "boxen": "1.0.0",
     "chalk": "1.1.3",
     "copy-paste": "1.3.0",
+    "dargs": "^5.1.0",
```
Comment (leo) [inline package.json]: Please pin this dependency! => `5.1.0`

**Comment by dar5hak** [comment]
Added docs. Please have a look at `readme.md`.

**Comment by dar5hak** [comment]
Also, a small problem. I couldn't get the `ignore` argument to work. How do you invoke it on the CLI?

```bash
serve --ignore readme.md package.json ./
```

This fails with the statement: `Specified directory doesn't exist!`

**Comment by leo** [comment]
@dar5hak `serve --ignore "readme.md,package.json"

**Comment by tunnckoCore** [comment]
@leo @dar5hak why use comma-separated values? Why not just accept a glob patterns and pass them to some fast matcher like `nanomatch` or `micromatch`. Nanomatch has strict support for only globs, without exglobs and etc. And it is Bash spec compliant.

**Comment by leo** [comment]
@tunnckoCore 

1. Not accepting comma-separated values and accepting glob values are two completely different things and aren't mutually exclusive.

2. If we're going to support glob values, you'll still need to support normal strings as well. Otherwise it won't be possible to ignore a single file.

3. If we're going to support glob values, we'll still use comma-separated values for that. Otherwise minimist will treat one glob value as the directory to serve. This means that we can just go on with this PR like I said [here](https://github.com/zeit/serve/pull/113#issuecomment-280906855).

😊

**Comment by dar5hak** [comment]
@leo There. Everything works now. :tada:

**Comment by leo** [comment]
@dar5hak Thanks a lot! 😊

**Pull request merged**